In [1]:
from imports import *


# Learning Pandas

## First Steps
Let's import the data:

In [3]:
nba = pd.read_csv("../data/nba_all_elo.csv")
type(nba)

pandas.core.frame.DataFrame

Some first information:

In [5]:
len(nba)

126314

In [6]:
nba.shape

(126314, 23)

Head of the data:


In [8]:
nba.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  ...  win_equiv  opp_id  opp_fran  opp_pts  \
0            0     TRH  Huskies  ...  40.294830     NYK    Knicks       68   
1            0     NYK   Knicks  ...  41.705170     TRH   Huskies       66   
2            0     CHS    Stags  ...  42.012257     NYK    Knicks       47   
3            0     NYK   Knicks  ...  40.692783     CHS     Stags       63   
4            0     DTF  Falcons  ...  38.864048     WSC  Capitols       50   

   opp_elo_i  opp_elo_n  game_location  game_result  forecast notes  
0  1300.0000  1306.7233              H            L  0.640065   NaN  
1  1300.0000  1293.2767              A            W  0.359935   NaN  
2  1306.7233  1297.0712              H            W  0.631101   NaN  
3  1300.0000  1309.6521              A            L  0.368899   NaN  
4  1300.0000  1320.3811              H            L  0.640065   NaN  

[5 rows x 23 columns]

In order to show all the columns:

In [10]:
pd.set_option("display.max.columns", None)
nba.head()

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes  
0  0.640065   NaN  
1  0.359935   NaN  
2  0.631101   NaN  
3  0.368899   NaN  
4  0.640065   NaN

While it’s practical to see all the columns, you probably won’t need six decimal places! Change it to two:



In [11]:
pd.set_option("display.precision", 2)
nba.head()


gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts    elo_i    elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.00  1293.28      40.29    NYK   
1            0     NYK   Knicks   68  1300.00  1306.72      41.71    TRH   
2            0     CHS    Stags   63  1300.00  1309.65      42.01    NYK   
3            0     NYK   Knicks   47  1306.72  1297.07      40.69    CHS   
4            0     DTF  Falcons   33  1300.00  1279.62      38.86    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68    1300.00    1306.72             H           L   
1   Huskies       66    1300.00    1293.28             A           W   
2    Knicks       47    1306.72    1297.07             H           W   
3     Stags       63    1300.00    1309.65             A           L   
4  Capitols       50    1300.00    1320.38             H           L   

   forecast notes  
0      0.64   NaN  
1      0.36   NaN  
2      0.63   NaN  
3      0.37   NaN  
4      0.64   NaN

The tail



In [12]:
nba.tail()

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts    elo_i    elo_n  \
126309         100            1     CLE  Cavaliers   82  1723.41  1704.39   
126310         102            1     GSW   Warriors  104  1809.98  1813.63   
126311         101            1     CLE  Cavaliers   91  1704.39  1700.74   
126312         102            1     CLE  Cavaliers   97  1700.74  1692.09   
126313         103            1     GSW   Warriors  105  1813.63  1822.29   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
126309      60.31    GSW   Warriors      103    1790.96    1809.98   
126310      68.01    CLE  Cavaliers       91    1704.39    1700.74   
126311      60.01    GSW   Warriors      104    1809.98    1813.63   
126312      59.29    GSW   Warriors      105    1813.63    1822.29   
126313      68.52    CLE  Cavaliers       97    1700.74    1692.09   

       game_location game_result  forecast notes  
126309             H           L      0.55   NaN  
126310             H           W      0.77   NaN  
126311             A           L      0.23   NaN  
126312             H           L      0.48   NaN  
126313             A           W      0.52   NaN

## Displaying Data Types
The first step in getting to know your data is
to discover the different data types it contains.

In [13]:
nba.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   gameorder      126314 non-null  int64  
 1   game_id        126314 non-null  object 
 2   lg_id          126314 non-null  object 
 3   _iscopy        126314 non-null  int64  
 4   year_id        126314 non-null  int64  
 5   date_game      126314 non-null  object 
 6   seasongame     126314 non-null  int64  
 7   is_playoffs    126314 non-null  int64  
 8   team_id        126314 non-null  object 
 9   fran_id        126314 non-null  object 
 10  pts            126314 non-null  int64  
 11  elo_i          126314 non-null  float64
 12  elo_n          126314 non-null  float64
 13  win_equiv      126314 non-null  float64
 14  opp_id         126314 non-null  object 
 15  opp_fran       126314 non-null  object 
 16  opp_pts        126314 non-null  int64  
 17  opp_elo_i      126314 non-nul

### Showing Basics Statistics:



In [15]:
nba.describe()

gameorder   _iscopy    year_id  seasongame  is_playoffs        pts  \
count  126314.00  126314.0  126314.00   126314.00    126314.00  126314.00   
mean    31579.00       0.5    1988.20       43.53         0.06     102.73   
std     18231.93       0.5      17.58       25.38         0.24      14.81   
min         1.00       0.0    1947.00        1.00         0.00       0.00   
25%     15790.00       0.0    1975.00       22.00         0.00      93.00   
50%     31579.00       0.5    1990.00       43.00         0.00     103.00   
75%     47368.00       1.0    2003.00       65.00         0.00     112.00   
max     63157.00       1.0    2015.00      108.00         1.00     186.00   

           elo_i      elo_n  win_equiv    opp_pts  opp_elo_i  opp_elo_n  \
count  126314.00  126314.00  126314.00  126314.00  126314.00  126314.00   
mean     1495.24    1495.24      41.71     102.73    1495.24    1495.24   
std       112.14     112.46      10.63      14.81     112.14     112.46   
min      1091.64    1085.77      10.15       0.00    1091.64    1085.77   
25%      1417.24    1416.99      34.10      93.00    1417.24    1416.99   
50%      1500.95    1500.95      42.11     103.00    1500.95    1500.95   
75%      1576.06    1576.29      49.64     112.00    1576.06    1576.29   
max      1853.10    1853.10      71.11     186.00    1853.10    1853.10   

        forecast  
count  126314.00  
mean        0.50  
std         0.22  
min         0.02  
25%         0.33  
50%         0.50  
75%         0.67  
max         0.98

In [16]:
nba.describe(include=object)


game_id   lg_id  date_game team_id fran_id  opp_id opp_fran  \
count         126314  126314     126314  126314  126314  126314   126314   
unique         63157       2      12426     104      53     104       53   
top     197302140MMT     NBA  4/13/2011     BOS  Lakers     BOS   Lakers   
freq               2  118016         30    5997    6024    5997     6024   

       game_location game_result           notes  
count         126314      126314            5424  
unique             3           2             231  
top                H           L  at New York NY  
freq           63138       63157             440

### Exploring Your Dataset
For example, you can examine how often specific values occur in a column:

In [17]:
nba["team_id"].value_counts()



BOS    5997
NYK    5769
LAL    5078
DET    4985
PHI    4533
       ... 
INJ      60
TRH      60
PIT      60
DTF      60
SDS      11
Name: team_id, Length: 104, dtype: int64

In [18]:
nba["fran_id"].value_counts()


Lakers          6024
Celtics         5997
Knicks          5769
Warriors        5657
Pistons         5650
Sixers          5644
Hawks           5572
Kings           5475
Wizards         4582
Spurs           4309
Bulls           4307
Pacers          4227
Thunder         4178
Rockets         4154
Nuggets         4120
Nets            4106
Suns            4080
Bucks           4034
Trailblazers    3870
Cavaliers       3810
Clippers        3733
Jazz            3555
Mavericks       3013
Heat            2371
Pelicans        2254
Magic           2207
Timberwolves    2131
Grizzlies       1657
Raptors         1634
Hornets          894
Colonels         846
Squires          799
Spirits          777
Stars            756
Sounds           697
Baltimore        467
Floridians       440
Condors          430
Capitols         291
Olympians        282
Sails            274
Stags            260
Bombers          249
Steamrollers     168
Packers           72
Redskins          65
Rebels            63
Denver       

It seems that a team named "Lakers" played 6024 games,
but only 5078 of those were played by the Los Angeles Lakers.
Find out who the other "Lakers" team is:

In [20]:
nba.loc[nba["fran_id"] == "Lakers", "team_id"].value_counts()


LAL    5078
MNL     946
Name: team_id, dtype: int64

## Getting to Know Pandas’ Data Structures
While a DataFrame provides functions that can feel quite intuitive,
the underlying concepts are a bit trickier to understand.
For this reason, you’ll set aside the vast NBA DataFrame and
build some smaller Pandas objects from scratch.

In [21]:
revenues = pd.Series([5555, 7000, 1980])
revenues


0    5555
1    7000
2    1980
dtype: int64

You’ve used the list [5555, 7000, 1980] to create a Series object called revenues. A Series object wraps two components:

A sequence of values
A sequence of identifiers, which is the index
You can access these components with .values and .index, respectively:


In [22]:
revenues.values

array([5555, 7000, 1980])

In [23]:
revenues.index


RangeIndex(start=0, stop=3, step=1)

While Pandas builds on NumPy, a significant difference is in
their indexing. Just like a NumPy array,
a Pandas Series also has an integer index that’s implicitly defined.
This implicit index indicates the element’s position in the Series.

However, a Series can also have an arbitrary type of index.
You can think of this explicit index as labels for a specific row:


In [24]:
city_revenues = pd.Series(
    [4200, 8000, 6500],
    index=["Amsterdam", "Toronto", "Tokyo"]
)
city_revenues


Amsterdam    4200
Toronto      8000
Tokyo        6500
dtype: int64

1. revenues: This Series behaves like a Python list because it only has a positional index.

1. city_revenues: This Series acts like a Python dictionary because it features both a positional and a label index.


In [25]:
city_employee_count = pd.Series({"Amsterdam": 5, "Tokyo": 8})
city_employee_count


Amsterdam    5
Tokyo        8
dtype: int64

In [26]:
city_employee_count.keys()

Index(['Amsterdam', 'Tokyo'], dtype='object')

In [27]:
"Tokyo" in city_employee_count

True

In [28]:
"New York" in city_employee_count


False